### ISSUE TITLE
Challenge for com/sun/ts/tests/ejb30/lite/xmloverride/resource/JsfClient.java\#unmappedEnvEntry_from_ejblitejsf

### ISSUE THREAD

**Issue by chengfang** [eclipse-ee4j/jakartaee-tck#1063]
1. The relevant specification version and section number(s). Jakarta EE Platform Spec 5.4.1.3. Declaration of Simple Environment Entries

3. The coordinates of the challenged test(s): com/sun/ts/tests/ejb30/lite/xmloverride/resource/JsfClient.java\#unmappedEnvEntry_from_ejblitejsf

5. The exact TCK version.  Jakarta EE TCK 10

7. The implementation being tested, including name and company. WildFly 27 Preview, Red Hat

9. A full description of why the test is invalid and what the correct behavior is believed to be.
10. Any supporting material; debug logs, test output (.jtr file), test logs, run scripts, etc.

The test class jsfclient contains a @Resource injection of env-entry, but there is no declaration of this env-entry in descriptors.

```java
public class JsfClient extends EJBLiteJsfClientBase {
  @Resource(name = "unmappedEnvEntry")
  private Integer unmappedEnvEntry = 1;

  private TestBean getTestBean() {
    return (TestBean) ServiceLocator
        .lookupNoTry("java:global/" + getModuleName() + "/TestBean");
  }

  /*
   * @testName: unmappedEnvEntry
   *
   * @test_Strategy: this @Resource has no mapped <env-entry> in descriptor, so
   * the default value of the field is unchanged.
   */
  public void unmappedEnvEntry() {
    assertEquals("Check correct the default value is retained. ", 1,
        unmappedEnvEntry);
  }
```
The spec says `env-entry-valuie` can be omitted so that the java default value can be used. But it does not say that the entire `<env-entry>` element can be missing. 

This test does not declare this `env-entry` in descriptor, nor does it have any `lookup` attribute to provide a value. So this `@Resource` injection is invalid.

5.4.1.3. Declaration of Simple Environment Entries:

>It’s often convenient to declare a field or method as an injection target, but specify a default value in the code, as illustrated in the following example.

```java
   // The maximum number of tax exemptions, configured by the Deployer.
  @Resource int maxExemptions = 4;        // defaults to 4
```
>To support this case, the container must only inject a value for this resource if the deployer has specified a value to override the default value. The env-entry-value element in the deployment descriptor is optional when an injection target is specified. If the element is not specified, no value will be injected. In addition, if the element is not specified, the named resource is not initialized in the naming context; explicit lookups of the named resource will fail.

**Comment by chengfang** [comment]
I agree that an env-entry can be defined with annotation alone, for example, the following annotation-defined env-entry are perfectly fine:

```java
@Resource(lookup = "...") boolean b;

@Resource(mappedName = "...") boolean b;
```

But annotation without supplying a target value (e.g., `@Resource() boolean b`) is an unresolvable resource reference. The only way to make this bare annotation work is to corroborate it with a corresponding <env-entry> descriptor element.  This descriptor element declares the intention to supply a configurable value to the annotated resource. This descriptor element may defer supplying the env-entry-value, in which case, the default java field value takes effect.

In the case of this test, the bare @Resource annotation does not have an env-entry element in web.xml, so the deployer is not aware of such intention, and therefore the deployer cannot configure the application properly. That's why I think the incomplete @Resource annotation must have a corresponding element in descriptor, with or without env-entry-value.

From a user's perspective, it's also better to fail such unresolvable resource reference. A bare @Resource annotation could be either:

(1) the user forgot to supply a value; or 
(2) the user wants the java default value for now, and may be configured to a different value later. How to configure it to a different value? Through descriptor.  So naturally it would serve the user best to have such an element in descriptor so that the user / deployer will know how to configure it, without looking at source code. Note that the source code may not be available when reconfiguring it.

I guess (1) is probably more common than (2). If the spec allows such unresolvable resource reference to go by silently, the user in (1) will not be alerted of this potential problem of missing configuration.

**Comment by tkburroughs** [comment]
Section 5.2.1 The Application Component's Environment states the following:

> The Deployer uses the tools provided by the container to initialize the environment entries that are
declared in the application component’s deployment descriptor or via annotations. The Deployer
can set and modify the values of the environment entries. As part of this process, the Deployer is
allowed to override any “lookup” information associated with the application component.

The intent is that the deployment tooling would see all `@Resource` annotations, not just resources defined in XML, and present those to the deployer to assign or change the value if desired. There should be no reason to specify any information in an XML file if the deployer does not wish to specify a value, but instead let the field default to the value assigned to it in the Java source code.

Also, there is no requirement that the "value" even be in web.xml/ejb-jar.xml.  The last part of the above sentence specifically calls out "lookup", but it could really override any information that indicates how to obtain the "value".  All Jakarta EE vendors may define "values" or binding information in vendor specific "binding" files.  This allows deployment tooling to specify a value in a separate file without modifying any of the application files.  For example, in Open Liberty, the "value" for such an `@Resoure` may be specified in an Open Liberty specific file called `server.xml`.

> (1) the user forgot to supply a value; or

The deployment tooling should be presenting all resources declared either using annotations or XML, not just XML; as described by the specification

> (2) the user wants the java default value for now, and may be configured to a different value later.

The deployment tooling should allow the deployer to specify a value via any mechanism the vendor supports.  There is no reason the "value" must be specified in the deployment descriptor; deployment tooling may instead specify the value in a vendor specific binding file (supported by many vendors).

From a user's perspective, there is no requirement to include either a web.xml or ejb-jar.xml.  All resource references and the bindings/values may be specified either using annotations, the deployment descriptor as an override of the annotations, or in vendor specific configuration.

This has been the behavior of `@Resource` since it was introduced in Java EE 5, when annotations in general were introduced, with the intent to allow Java EE applications without the need for a deployment descriptor.

**Comment by scottmarlow** [comment]
Closing this as `invalid` based on the above feedback.
